### PR TITLE
Add documentation for Base.Order module

### LINKED ISSUE #18370: No documentation for `Base.Order` and `Ordering` for sort and search functions
Julia 0.5-rc3 contains new methods for `searchsortedfirst()` and many other functions that involve specifying the sort pattern using singleton types defined in `Base.Order`. This is really awesome, and I like it a lot, but the users should probably be able to find out about this in the documentation and docstrings.

At the very least, I feel that a NEWS.md entry should be added before tagging v0.5 final, so that new v0.5 users may know about these changes. It certainly surprised me in #18369 :)

### PR DESCRIPTION
**Pull request by jlumpe**
Adds docstrings to `Base.Order` module and adds a section to the "Sorting and Related Functions" documentation page. Resolves #18370.

### CODE REVIEW

**Comment by jlumpe** [comment]
I did not add a docstring to the `ordtype` function because I can't actually tell what it does, but it seems like it is meant to be deprecated in the future anyway.

**Comment by jlumpe** [comment]
Any chance of getting this reviewed?

**Comment by jlumpe** [comment]
Bump?

File: base/ordering.jl [hunk 1]
```diff
@@ -18,9 +18,24 @@ export # not exported by Base
     DirectOrdering,
     lt, ord, ordtype
 
+"""
```
Comment (nalimilan) [inline base/ordering.jl:21]: Better show the fully qualified name as it's not exported:
```suggestion
"""
        Base.Order.Ordering

```

Same elsewhere.

File: base/ordering.jl [hunk 2]
```diff
@@ -18,9 +18,24 @@ export # not exported by Base
...
 abstract type Ordering end
 
 struct ForwardOrdering <: Ordering end
+
+"""
+    ReverseOrdering(fwd::Ordering=Forward)
+
+A wrapper which reverses an ordering.
+
+For a given `o::Ordering`, the following holds for all  `a`, `b`:
```
Comment (nalimilan) [inline base/ordering.jl:35]:
```suggestion
For a given `Ordering` `o`, the following holds for all  `a`, `b`:
```

File: base/ordering.jl [hunk 3]
```diff
@@ -31,9 +46,21 @@ ReverseOrdering() = ReverseOrdering(ForwardOrdering())
 
 const DirectOrdering = Union{ForwardOrdering,ReverseOrdering{ForwardOrdering}}
 
+"""
```
Comment (nalimilan) [inline base/ordering.jl:49]:
```suggestion
"""
    Base.Order.Forward

```

File: base/ordering.jl [hunk 4]
```diff
@@ -31,9 +46,21 @@ ReverseOrdering() = ReverseOrdering(ForwardOrdering())
...
+Default ordering according to [`isless`](@ref).
+"""
 const Forward = ForwardOrdering()
+
+"""
```
Comment (nalimilan) [inline base/ordering.jl:54]:
```suggestion
"""
    Base.Order.Reverse

```

File: base/ordering.jl [hunk 5]
```diff
@@ -31,9 +46,21 @@ ReverseOrdering() = ReverseOrdering(ForwardOrdering())
...
+Reverse ordering according to [`isless`](@ref).
+"""
 const Reverse = ReverseOrdering()
 
+"""
+    By(by, order::Ordering=Forward)
+
+Applies `order` to elements after they have been transformed by the function `by`.
```
Comment (nalimilan) [inline base/ordering.jl:62]:
```suggestion
`Ordering` which applies `order` to elements after they have been transformed
by the function `by`.
```

File: base/ordering.jl [hunk 6]
```diff
@@ -42,10 +69,23 @@ end
 # backwards compatibility with VERSION < v"1.5-"
 By(by) = By(by, Forward)
 
+"""
+    Lt(lt)
+
+Ordering which uses the function `lt(a, b)` to compare elements. `lt` should
```
Comment (nalimilan) [inline base/ordering.jl:75]:
```suggestion
`Ordering` which calls `lt(a, b)` to compare elements. `lt` should
```

File: base/ordering.jl [hunk 7]
```diff
@@ -42,10 +69,23 @@ end
...
+obey the same rules as implementations of [`isless`](@ref).
+"""
 struct Lt{T} <: Ordering
     lt::T
 end
 
+"""
+    Perm(order::Ordering, data::AbstractVector)
+
+Ordering on the indices of `data` where `i` is less than `j` if `data[i]` is
```
Comment (nalimilan) [inline base/ordering.jl:85]:
```suggestion
`Ordering` on the indices of `data` where `i` is less than `j` if `data[i]` is
```

File: base/ordering.jl [hunk 8]
```diff
@@ -54,6 +94,11 @@ end
 ReverseOrdering(by::By) = By(by.by, ReverseOrdering(by.order))
 ReverseOrdering(perm::Perm) = Perm(ReverseOrdering(perm.order), perm.data)
 
+"""
+    lt(o::Ordering, a, b)
+
+Test whether `a` is less than `b` according to the order `o`.
```
Comment (nalimilan) [inline base/ordering.jl:100]:
```suggestion
Test whether `a` is less than `b` according to the ordering `o`.
```

**Review by nalimilan — COMMENTED** [review]
Thanks for this! Looks good apart from a few details.

**Comment by jlumpe** [comment]
Thanks for the review! I have committed your suggestions.

**Comment by StefanKarpinski** [comment]
I've restarted the two failing CI runs, but it seems fairly unlikely that this change could possibly be causing failures.

**Comment by jlumpe** [comment]
It seems that one succeeded but the other is still failing. I'm not sure how to interpret the output of the failed one, but I agree this doesn't seem like it should break anything (especially garbage collection, which is what I think this test is?).

**Comment by StefanKarpinski** [comment]
I'm going to assume it can't possibly be relevant.

**Pull request merged**